### ISSUE TITLE
Support for async-streams (C# 8.0)

### ISSUE THREAD

**Issue by jcouv** [mono/mono#12019]
The types should be locked down in Core by mid-January.

The core new types the compiler needs are:
- [x] IAsyncEnumerable<T>
- [x] IAsyncEnumerator<T>
- [x] IAsyncDisposable
- [x] ManualResetValueTaskSourceCore
- [x] AsyncIteratorStateMachineAttribute

Additional fixes to be ported
~- [ ] https://github.com/dotnet/coreclr/pull/21103 (pretty-printing of stack traces)~ is not needed as we don't prettify iterators

Tagging @stephentoub @marek-safar 

Relates to https://github.com/dotnet/roslyn/issues/24037 (umbrella issue for async-streams)

**Comment by baulig** [comment]
The only type that's creating problems is `ManualResetValueTaskSourceCore`:

```
/Workspace/mono-lavender/external/corefx/src/Common/src/CoreLib/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs(167,79): error CS1739: The best overload for 'QueueUserWorkItem' does not have a parameter named 'preferLocal'
/Workspace/mono-lavender/external/corefx/src/Common/src/CoreLib/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs(171,85): error CS1739: The best overload for 'UnsafeQueueUserWorkItem' does not have a parameter named 'preferLocal'
/Workspace/mono-lavender/external/corefx/src/Common/src/CoreLib/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs(215,25): error CS1661: Cannot convert lambda expression to delegate type 'ContextCallback' because the parameter types do not match the delegate parameter types
/Workspace/mono-lavender/external/corefx/src/Common/src/CoreLib/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs(215,70): error CS1678: Parameter 1 is declared as type 'ref System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<TResult>' but should be 'object'
/Workspace/mono-lavender/external/corefx/src/Common/src/CoreLib/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs(216,29): error CS1615: Argument 3 may not be passed with the 'ref' keyword
/Workspace/mono-lavender/external/corefx/src/Common/src/CoreLib/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs(239,93): error CS1739: The best overload for 'QueueUserWorkItem' does not have a parameter named 'preferLocal'
/Workspace/mono-lavender/external/corefx/src/Common/src/CoreLib/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs(243,99): error CS1739: The best overload for 'UnsafeQueueUserWorkItem' does not have a parameter named 'preferLocal'
make[1]: *** [../../build/library.make:335: ../../class/lib/net_4_x-macos/mscorlib.dll] Error 1
```

We're still using `ThreadPool` from the reference source.

**Comment by baulig** [comment]
Experimental PR for this: https://github.com/mono/mono/pull/12674

**Comment by baulig** [comment]
I implemented all the types, how do I test this?

**Comment by stephentoub** [comment]
You can use C# 8 from VS2019 Preview 2 and write something like:
```C#
async IAsyncEnumerable<int> ProduceNumbersAsync()
{
    for (int i = 0; i < 10; i++) { await Task.Delay(1); yield return i * 2; }
}

static async Task Main()
{
    await foreach (int i in ProduceNumbersAsync()) Console.WriteLine(i);
}
```

**Comment by baulig** [comment]
All the types have landed, but still need to do the stack trace pretty-print thing.

**Comment by jcouv** [comment]
Awesome. Thanks!

**Comment by marek-safar** [comment]
@baulig could you land the stack trace change?

**Comment by jcouv** [comment]
Thanks!
Out of curiosity, is there a shipping vehicle that we can announce that will contain these new types?

**Comment by marek-safar** [comment]
@jcouv they will be included in Mono 6.2
